usersession/client/client.go

```diff
@@ -0,0 +1,244 @@
+	return net.Dial("unix", socket)
+}
+
+type Client struct {
+	doer *http.Client
+}
+
+func New() *Client {
+	transport := &http.Transport{Dial: dialSessionAgent, DisableKeepAlives: true}
+	return &Client{
+		doer: &http.Client{Transport: transport},
+	}
+}
+
+type Error struct {
+	Kind    string      `json:"kind"`
+	Value   interface{} `json:"value"`
+	Message string      `json:"message"`
+}
+
+func (e *Error) Error() string {
+	return e.Message
+}
+
+type response struct {
+	// Not from JSON
+	uid        int
+	err        error
+	statusCode int
+
+	Result json.RawMessage `json:"result"`
+	Type   string          `json:"type"`
+}
+
+func (resp *response) checkError() {
+	if resp.Type != "error" {
+		return
+	}
+	var resultErr Error
+	err := json.Unmarshal(resp.Result, &resultErr)
+	if err != nil || resultErr.Message == "" {
+		resp.err = fmt.Errorf("server error: %q", http.StatusText(resp.statusCode))
+	} else {
+		resp.err = &resultErr
+	}
+}
+
+func (client *Client) do(ctx context.Context, method, urlpath string, query url.Values, headers map[string]string, body []byte) ([]*response, error) {
+	sockets, err := filepath.Glob(filepath.Join(dirs.XdgRuntimeDirGlob, "snapd-session-agent.socket"))
+	if err != nil {
+		return nil, err
+	}
+	var (
+		wg        sync.WaitGroup
+		mu        sync.Mutex
+		responses []*response
+	)
+	for _, socket := range sockets {
+		wg.Add(1)
+		go func(socket string) {
+			defer wg.Done()
+			uidStr := filepath.Base(filepath.Dir(socket))
+			uid, err := strconv.Atoi(uidStr)
+			if err != nil {
+				logger.Noticef("Socket %q does not appear to be in a valid XDG_RUNTIME_DIR", socket)
+				return
+			}
+			u := url.URL{
+				Scheme:   "http",
+				Host:     uidStr,
+				Path:     urlpath,
+				RawQuery: query.Encode(),
+			}
+			req, err := http.NewRequest(method, u.String(), bytes.NewBuffer(body))
+			if err != nil {
+				logger.Noticef("Failed to create HTTP request: %v", err)
+				return
+			}
+			req = req.WithContext(ctx)
+			for key, value := range headers {
+				req.Header.Set(key, value)
+			}
+			httpResp, err := client.doer.Do(req)
+			response := response{
+				uid: uid,
+				err: err,
+			}
+			if err == nil {
+				defer httpResp.Body.Close()
+				response.statusCode = httpResp.StatusCode
+				response.err = decodeInto(httpResp.Body, &response)
+				response.checkError()
+			}
+			mu.Lock()
+			defer mu.Unlock()
+			responses = append(responses, &response)
+		}(socket)
+	}
+	wg.Wait()
+	return responses, nil
+}
+
+func decodeInto(reader io.Reader, v interface{}) error {
+	dec := json.NewDecoder(reader)
+	if err := dec.Decode(v); err != nil {
+		r := dec.Buffered()
+		buf, err1 := ioutil.ReadAll(r)
+		if err1 != nil {
+			buf = []byte(fmt.Sprintf("error reading buffered response body: %s", err1))
+		}
+		return fmt.Errorf("cannot decode %q: %s", buf, err)
+	}
+	return nil
+}
+
+type SessionInfo struct {
+	Version string `json:"version"`
+}
+
```

Comment:
Thinking about this request for a second. I think we're missing a RestartSelf API so that client daemon can restart itself after new snapd update. This is not strictly related to this PR but I think we need to address this in general before we start using the daemon -> per user daemon APIs more.

Reply:
It would be a nice to have, but note the session agent goes away on its own after 30s